File: kryptografia/cryptography/macierzowe.py
# -*- coding: utf-8 -*-

__author__ = "Rafał Selewońko <rafal@selewonko.com>"


class MacierzoweA(object):
    text = ''
    key = []
    sortedKey = []
    stringArray = []

    def __init__(self):
        pass

    def setText(self, s):
        self.text = s

    def setKey(self, s):
        self.key = []
        nums = s.split("-")
        for n in nums:
            self.key.append(int(n))
        self.sortedKey = self.key[:]
        self.sortedKey.sort()

    def reverseKey(self):
        self.key.reverse()

    def czyCiagla(self):
        ret = True
        for i in range(self.maxKey()):
            if i + 1 not in self.key:
                ret = False
        return ret

    def sprawdzKlucz(self):
        ret = self.czyCiagla()

        for i in range(len(self.sortedKey)):
            if i == 0:
                tmp = self.sortedKey[i]
            if i > 0 and i < len(self.sortedKey) - 1:
                tmp = self.sortedKey[i]
                if tmp == self.sortedKey[i + 1]:
                    ret = False
        return ret

    def maxKey(self):
        return max(self.key)

    def countRows(self):
        ret = len(self.text) / self.maxKey()
        if len(self.text) % self.maxKey() > 0:
            ret += 1
        return ret

    def makeArray(self):
        i = 1
        max = self.maxKey()
        while len(self.stringArray) != self.countRows():
            first = i * max - max
            second = i * max
            if second > len(self.text):
                second = len(self.text)
            self.stringArray.append(self.text[first:second])
            i += 1

    def encrypted(self):
        self.makeArray()
        ret = ""
        for s in self.stringArray:
            for i in self.key:
                if len(s) > i - 1:
                    ret += s[i - 1]
        return ret

    def decrypted(self):
        self.makeArray()
        ret = ''

        self.reverseKey()
        for s in self.stringArray:
            licznik = 0
            m = len(s)
            for i in self.key:
                if licznik < m:
                    j = i - 1
                    while (j >= m):
                        j -= 1
                    ret += s[j]
                    licznik += 1
        return ret

    def encrypt(self, string, key):
        self.clear()
        self.setText(string)
        self.setKey(key)
        if not self.sprawdzKlucz():
            return "klucz nieciagly"
        else:
            return self.encrypted()

    def decrypt(self, string, key):
        self.clear()
        self.setText(string)
        self.setKey(key)
        if not self.sprawdzKlucz():
            return "klucz nieciagly"
        else:
            return self.decrypted()

    def clear(self):
        self.text = ""
        self.key = []
        self.sortedKey = []
        self.stringArray = []

File: kryptografia/cryptography/macierzowe.java
# -*- coding: utf-8 -*-

__author__ = "Rafał Selewońko <rafal@selewonko.com>"


class MacierzoweA(object):
    private static String text;
    private static ArrayList<Integer> key         = new ArrayList<Integer>();
    private static ArrayList<Integer> sortedKey   = new ArrayList<Integer>();
    private static ArrayList<String>  stringArray = new ArrayList<String>();

    public static void getText(String in){ text = in; }
    public static void getKey(String in){
         key.clear();
         String nums[] = in.split("-");
         for(String n : nums) key.add(Integer.parseInt(n));
         sortedKey = (ArrayList)key.clone();
         Collections.sort(sortedKey);
    }
    private static void reverseKey(){ Collections.reverse(key); }
    public static void showKey(){
        System.out.println("\nKlucz: ");
        for(Integer i : key) System.out.print(i+" ");
        System.out.println("\nPosortowany Klucz: ");
        for(Integer i : sortedKey) System.out.print(i+" ");
        
        System.out.print("\nMax: " + maxKey());
        System.out.print("\nCiagla: " + czyCiagla() + "\n");
    } 
    private static boolean czyCiagla(){
        boolean ret = true;
        for(int i=1;i<maxKey();i++)
            if(!key.contains(i)) ret= false;
        return ret;
    }
    private static boolean sprawdzKlucz(){
        boolean ret   = czyCiagla();
        boolean start = true;
        Integer tmp;
        
        for(int i=0;i<sortedKey.size();i++){
            if(i==0){
                tmp=sortedKey.get(i);
            }
            if((i>0)&&(i<sortedKey.size()-1)){
                tmp=sortedKey.get(i);
                if(tmp == sortedKey.get(i+1)) ret=false;
            }
        }
        return ret;
    }
    private static Integer maxKey(){ return Collections.max(key); }
    private static Integer countRows(){
        int ret = (text.length() / maxKey());
        if( (text.length()%maxKey()) > 0 ) ret += 1;
        return ret;
    }
    private static void makeArray(){
        int i   = 1;
        int max = maxKey();
        while(stringArray.size()!=countRows()){
            int first  = i*max - max;
            int second = i*max;
            if(second > text.length()) second = text.length();
            stringArray.add( text.substring(first, second) );
            i++;
        }
    }
    public static void showArray(){
        makeArray();
        for(Integer i : sortedKey) System.out.print(i);
        System.out.println("\n------------");
        for(String s: stringArray) System.out.println(s);
    }
    public static String Encrypted(){
        makeArray();
        String ret = "";
        for(String s : stringArray){
            char[] arr = s.toCharArray();
            for(Integer i : key){
                 if(arr.length > i-1)
                     ret += arr[i-1];
            }
        }
        return ret;
    }
    public static String Decrypted(){
        makeArray();
        String ret = "";
        reverseKey();
        for(String s : stringArray){
            char[] arr = s.toCharArray();
                if(arr.length > maxKey()-1){
                    for(Integer i : key){
                        //if(arr.length > i-1)
                            ret += arr[i-1];
                    }
                }
                else
                { 
                    int licznik=0;
                    int m = arr.length;
                    for(Integer i : key){
                        if(licznik<m){
                            int j = i-1;
                            while(j >= m) j--;
                            ret += arr[j];
                            licznik++;
                        }
                    }
                }
        }
        return ret;
    }
    
    public static String Encrypt(){
        if(sprawdzKlucz()==false) return "Blad ! Klucz nie jest ciagly albo zawiera duplikaty.\n";
        else return Encrypted();
    }
    public static String Decrypt(){
        if(sprawdzKlucz()==false) return "Blad ! Klucz nie jest ciagly albo zawiera duplikaty.\n";
        else return Decrypted();
    }
    
    public static void clear(){
        text = new String();
        key = new ArrayList<Integer>();
        sortedKey = new ArrayList<Integer>();
        stringArray = new ArrayList<String>();
    }
}
